Computer Architecture

VeraSimon opened the pull request:
### Day 1: Get `print8.ls8` running

- [x] Inventory what is here
- [x] Implement `struct cpu` in `cpu.h`
- [x] Add RAM functions `cpu_ram_read` and `cpu_ram_write`
- [x] Implement `cpu_init()`
- [x] Implement the core of `cpu_run()`
- [x] Implement the `HLT` instruction handler
- [x] Add the `LDI` instruction
- [x] Add the `PRN` instruction

### Day 2: Add the ability to load files dynamically, get `mult.ls8` and `stack.ls8` running

- [x] Un-hardcode the machine code
- [x] Implement the `cpu_load` function to load an `.ls8` file given the
      filename passed in as an argument
- [x] Implement a Multiply instruction and Print the result (run `mult8.ls8`)

### Day 3; Stack

- [x] Implement the System Stack and be able to run the `stack.ls8` program

### Day 4: Get `call.ls8` running

- [x] Implement the CALL and RET instructions
- [x] Implement Subroutine Calls and be able to run the `call.ls8` program

### Stretch

- [ ] Add the timer interrupt to the LS-8 emulator
- [ ] Add the keyboard interrupt to the LS-8 emulator
- [ ] Write an LS-8 assembly program to draw a curved histogram on the screen

VeraSimon committed:
gitignore

VeraSimon committed:
CPU
Implemented CPU struct

VeraSimon committed:
CPU
Updated CPU struct after a bit more reading of the docs

VeraSimon committed:
cpu_init()
Implemented cpu_init()

VeraSimon committed:
cpu_run()
Implemented some of core of cpu_run(), the HLT instruction, and a default error case

VeraSimon committed:
cpu_run()
Fleshed out the core of cpu_run() some more
Implemented LDI
Tweaked cpu.h so register and ram sizes were macro constants for use across the app

LuisanSuarez reviewed (approved):
.

LuisanSuarez commented on ls8/cpu.c:
```diff
@@ -4,21 +4,27 @@
  */
-// cpu_ram_read()
-int *cpu_ram_read(struct cpu *cpu)
+int cpu_ram_read(struct cpu *cpu, int address)
```
by far the most sophisticated verison of `cpu_ram_read()` I've seen so far. 3 stars

LuisanSuarez commented on ls8/cpu.c:
```diff
@@ -4,21 +4,27 @@
 
-// cpu_ram_write()
-int *cpu_ram_write(struct cpu *cpu)
+void cpu_ram_write(struct cpu *cpu, int address, int value)
```
same as above

LuisanSuarez commented on ls8/cpu.c:
```diff
@@ -69,35 +76,63 @@ void cpu_run(struct cpu *cpu)
     {
-        // TODO
+        // Debugging/testing
+        printf("Registers:\n[ ");
```
not to toot my own horn, but: great idea, will make debugging 10x easier

VeraSimon replied:
If it ain't broke...

LuisanSuarez commented on ls8/cpu.c:
```diff
@@ -69,35 +76,63 @@ void cpu_run(struct cpu *cpu)
+        switch (operands)
+        {
+        case 1:
+            operandA = cpu_ram_read(cpu, IR + 1);
```
very nice. keeping this in variable names will make it much easier to read

LuisanSuarez commented on ls8/cpu.c:
```diff
@@ -64,17 +66,38 @@ void alu(struct cpu *cpu, enum alu_op op, unsigned char regA, unsigned char regB
         // 6. Move the PC to the next instruction.
+        switch (IR)
+        {
+        case HLT:
```
simple

LuisanSuarez commented on ls8/cpu.c:
```diff
@@ -64,17 +66,38 @@ void alu(struct cpu *cpu, enum alu_op op, unsigned char regA, unsigned char regB
+            break;
+        case PRN:
+            break;
+        default:
```
nice

LuisanSuarez commented on ls8/cpu.c:
```diff
@@ -67,4 +84,13 @@ void cpu_run(struct cpu *cpu)
 void cpu_init(struct cpu *cpu)
```
👍

VeraSimon committed:
cpu_load()
Implemented cpu_load(), and verified its reading data into "ram" successfully

VeraSimon committed:
cpu_run()
Worked out the kinks from yesterday. print8.ls8 is now running as expected.

VeraSimon committed:
Misc
Cleaned up code a bit
Started implementing MUL instruction

VeraSimon committed:
MUL
Finished implementing MUL instruction

VeraSimon committed:
PUSH and POP
Implemented PUSH and POP instructions

VeraSimon committed:
Debugging
Fleshed out and cleaned up debugging output a bit

VeraSimon committed:
Refactor
Removed use of extra counter in cpu_run() and changed cpu->PC to an unsigned int in the process to streamline things

VeraSimon committed:
Instruction register
Moved the instruction register into the cpu struct. Refactored code using it accordingly.

VeraSimon committed:
ALU instructions
Implemented the remaining 4 listed ALU instructions because ADD was needed for call.ls8, and why not

VeraSimon committed:
Stack functions
Moved logic from PUSH and POP instructions into separate functions for easier use in CALL

LuisanSuarez commented on ls8/cpu.c:
```diff
@@ -26,18 +26,41 @@ void cpu_ram_write(struct cpu *cpu, int address, unsigned char value)
     }
 }
 
+void cpu_stack_push(struct cpu *cpu, int reg)
```
very fancy these two

LuisanSuarez commented on ls8/cpu.c:
```diff
@@ -75,11 +75,45 @@ void alu(struct cpu *cpu, enum alu_op op, unsigned char regA, unsigned char regB
 {
     switch (op)
     {
+    case ALU_ADD:
```
looks good :)

VeraSimon committed:
Refactoring
Refactored debugging so individual cpu functions didnt need the flag as part of their initialization
Refactored cpu_stack_pop to return the popped value
CALL & RET still WIP

VeraSimon committed:
CALL/RET
Implemented CALL & RET

VeraSimon committed:
Program counter
Streamlined pointer counter incrementation code for future expansion of project

VeraSimon committed:
Cleanup
Removed a thing from cpu.h that wasnt needed
